packages/beacon-node/src/chain/chain.ts

```diff
@@ -333,11 +333,22 @@ export class BeaconChain implements IBeaconChain {
     return headState;
   }
 
-  async getHeadStateAtCurrentEpoch(): Promise<CachedBeaconStateAllForks> {
-    const currentEpochStartSlot = computeStartSlotAtEpoch(this.clock.currentEpoch);
+  async getHeadStateAtCurrentEpoch(regenCaller: RegenCaller): Promise<CachedBeaconStateAllForks> {
+    return this.getHeadStateAtEpoch(this.clock.currentEpoch, regenCaller);
+  }
+
+  async getHeadStateAtEpoch(epoch: Epoch, regenCaller: RegenCaller): Promise<CachedBeaconStateAllForks> {
+    // using getHeadState() means we'll use checkpointStateCache if it's available
+    const headState = this.getHeadState();
+    // head state is in the same epoch, or we pulled up head state already from past epoch
+    if (epoch <= computeEpochAtSlot(headState.slot)) {
```

Comment:
@tuyennhv shouldn't this be a strict equals check? I noticed this method always returns the same state if a previous epoch is passed in

Reply:
@nflaig no this should be "<=", the function means to get head state dialing to an epoch. If provided epoch < head state epoch just return head state

also if we let it pass through the below `regen.getBlockSlotState` call, if provided slot < head state slot it'll throw error

Reply:
That makes sense, I was trying to figure out what would be required to get proposer duties of a finalized epoch. The right method to use in that case would be `getStateBySlot` I'd assume. 
https://github.com/ChainSafe/lodestar/blob/f77b774be201b30a4451435ad90a48c5dee48029/packages/beacon-node/src/chain/chain.ts#L362
 You also noted in https://github.com/ChainSafe/lodestar/issues/5846 is that Lodestar does not persists finalized states very frequently which would be an issue in that case. We could solve this issue if we add `--chain.archiveStateEpochFrequency 1` flag but this significantly increases storage requirements (but that might be fine for someone that really needs that data).